addons/src/ao_doclist.c

```diff
@@ -120,23 +124,60 @@ static void ao_doc_list_finalize(GObject *object)
 /* This function is taken from Midori's katze-utils.c, thanks to Christian. */
 static void ao_popup_position_menu(GtkMenu *menu, gint *x, gint *y, gboolean *push_in, gpointer data)
 {
+	AoDocListPrivate *priv = AO_DOC_LIST_GET_PRIVATE(data);
+
 	gint wx, wy;
 	GtkRequisition widget_req;
-	GtkWidget *widget = data;
-	GdkWindow *window = gtk_widget_get_window(widget);
+	GtkWidget *widget = GTK_WIDGET(priv->toolbar_doclist_button);
+	GdkWindow *window;
 	gint widget_height;
 
-	/* Retrieve size and position of both widget and menu */
-	if (! gtk_widget_get_has_window(widget))
-	{
-		GtkAllocation allocation;
-		gdk_window_get_position(window, &wx, &wy);
-		gtk_widget_get_allocation(widget, &allocation);
-		wx += allocation.x;
-		wy += allocation.y;
+	if (priv->in_overflow_menu) {
+		/* The button was added to the toolbar overflow menu (since the toolbar
+		 * isn't wide enough to contain the button), so use the toolbar window
+		 * instead
+		 */
+
+		/* Get the root coordinates of the toolbar's window */
+		int wx_root, wy_root, width;
+		window = gtk_widget_get_window (gtk_widget_get_ancestor(widget, GTK_TYPE_TOOLBAR));
+#if GTK_CHECK_VERSION(3, 0, 0)
+		gdk_window_get_geometry(window, &wx, &wy, &width, NULL);
+#else
+		gdk_window_get_geometry(window, &wx, &wy, &width, NULL, NULL);
+#endif
+		gdk_window_get_root_coords(window, wx, wy, &wx_root, &wy_root);
```

Comment:
Not completely sure but can't we move the `gdk_window_get_root_coords` call below the others and so merge the two GTK version check blocks?

Reply:
Yes, I will change this.